helper/convert/value_test.go

```diff
@@ -0,0 +1,420 @@
+					"weight": cty.NumberIntVal(20),
+					"toys":   cty.ListVal([]cty.Value{cty.StringVal("ball"), cty.StringVal("rope")}),
+				}),
+				cty.ObjectVal(map[string]cty.Value{
+					"name":   cty.StringVal("Bobby"),
+					"breed":  cty.StringVal("Golden"),
+					"weight": cty.NumberIntVal(30),
+					"toys":   cty.ListVal([]cty.Value{cty.StringVal("dummy"), cty.StringVal("frisbee")}),
+				}),
+			}),
+			Want: tftypes.NewValue(tftypes.List{
+				ElementType: tftypes.Object{
+					AttributeTypes: map[string]tftypes.Type{
+						"name":   tftypes.String,
+						"breed":  tftypes.String,
+						"weight": tftypes.Number,
+						"toys":   tftypes.List{ElementType: tftypes.String},
+					},
+				},
+			}, []tftypes.Value{
+				tftypes.NewValue(tftypes.Object{
+					AttributeTypes: map[string]tftypes.Type{
+						"name":   tftypes.String,
+						"breed":  tftypes.String,
+						"weight": tftypes.Number,
+						"toys":   tftypes.List{ElementType: tftypes.String},
+					},
+				}, map[string]tftypes.Value{
+					"name":   tftypes.NewValue(tftypes.String, "Alice"),
+					"breed":  tftypes.NewValue(tftypes.String, "Beagle"),
+					"weight": tftypes.NewValue(tftypes.Number, 20),
+					"toys": tftypes.NewValue(tftypes.List{ElementType: tftypes.String}, []tftypes.Value{
+						tftypes.NewValue(tftypes.String, "ball"),
+						tftypes.NewValue(tftypes.String, "rope"),
+					}),
+				}),
+				tftypes.NewValue(tftypes.Object{
+					AttributeTypes: map[string]tftypes.Type{
+						"name":   tftypes.String,
+						"breed":  tftypes.String,
+						"weight": tftypes.Number,
+						"toys":   tftypes.List{ElementType: tftypes.String},
+					},
+				}, map[string]tftypes.Value{
+					"name":   tftypes.NewValue(tftypes.String, "Bobby"),
+					"breed":  tftypes.NewValue(tftypes.String, "Golden"),
+					"weight": tftypes.NewValue(tftypes.Number, 30),
+					"toys": tftypes.NewValue(tftypes.List{ElementType: tftypes.String}, []tftypes.Value{
+						tftypes.NewValue(tftypes.String, "dummy"),
+						tftypes.NewValue(tftypes.String, "frisbee"),
+					}),
+				}),
+			}),
+		},
+		{
+			Value: cty.ListVal([]cty.Value{
+				cty.ObjectVal(map[string]cty.Value{
+					"enforcement": cty.NullVal(cty.String),
+				}),
+			}),
+			Want: tftypes.NewValue(tftypes.List{
+				ElementType: tftypes.Object{
+					AttributeTypes: map[string]tftypes.Type{
+						"enforcement": tftypes.String,
+					},
+				},
+			}, []tftypes.Value{
+				tftypes.NewValue(tftypes.Object{
+					AttributeTypes: map[string]tftypes.Type{
+						"enforcement": tftypes.String,
+					},
+				}, map[string]tftypes.Value{
+					"enforcement": tftypes.NewValue(tftypes.String, nil),
+				}),
+			}),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.Value.GoString(), func(t *testing.T) {
+			t.Parallel()
+
+			got := ListTfValue(test.Value)
+
+			if diff := cmp.Diff(test.Want, got); diff != "" {
+				t.Errorf("unexpected differences: %s", diff)
+			}
+		})
+	}
+}
+
+func TestSetTfType(t *testing.T) {
```

Comment:
> Same comment for the map + tuple tests

nit: I think it'd be nice to do what you did for the list test, which is add a test that goes through at least one level of recursion (a set of objects with a set attribute)

Reply:
Totally forgot maps 🤦‍♀️